lib/_stream_readable.js

```diff
@@ -1070,13 +1072,90 @@ Readable.prototype.wrap = function(stream) {
 };
 
 Readable.prototype[SymbolAsyncIterator] = function() {
-  if (createReadableStreamAsyncIterator === undefined) {
-    createReadableStreamAsyncIterator =
-      require('internal/streams/async_iterator');
+  let stream = this;
+
+  if (typeof stream.read !== 'function') {
+    // v1 stream
+    const src = stream;
+    stream = new Readable({
+      objectMode: true,
+      destroy(err, callback) {
+        destroyImpl.destroyer(src, err);
+        callback();
+      }
+    }).wrap(src);
   }
-  return createReadableStreamAsyncIterator(this);
+
+  const iter = createAsyncIterator(stream);
+  iter.stream = stream;
+  return iter;
 };
 
+async function* createAsyncIterator(stream) {
+  const state = stream._readableState;
+
+  if ((state && state.endEmitted) ||
+      (state && state.closeEmitted)) {
+    // TODO (ronag): What if errorEmitted?
+    // TODO (ronag): Wait for closeEmitted?
+    // TODO (ronag): if (closeEmitted && !endEmitted) ERR_PREMATURE_CLOSE?
+    stream.on('error', nop);
+    destroyImpl.destroyer(stream, null);
+    return;
+  }
+
+  let callback = nop;
+  let error;
+  let ended = false;
+  let closed = false;
+  let errored = false;
+
+  stream
+    .on('readable', () => {
+      callback();
+      callback = nop;
+    })
+    .on('error', (err) => {
+      errored = true;
+      error = err;
+      callback();
+      callback = nop;
+    })
+    .on('end', () => {
+      ended = true;
+      callback();
+      callback = nop;
+    })
+    .on('close', () => {
+      closed = true;
+      // TODO (ronag): ERR_PREMATURE_CLOSE?
+      callback();
+      callback = nop;
+    });
+
+  try {
+    while (!closed && !errored && !ended) {
+      const chunk = stream.read();
+      if (chunk !== null) {
+        yield chunk;
+      } else {
+        await new Promise((resolve) => {
+          callback = resolve;
+        });
```

Comment:
Can you move this closure outside of the try-catch? There is no need to allocate a new one for every chunk.

Reply:
I could move it out of the try/catch, but it still needs to live in a try/finally or the destroyer won't always be invoked.

Reply:
The current code allocates one closure for each iteration of the loop that needs to be awaited. If you move this outside of the loop, you can save a significant amount of closure allocations.

Reply:
It gets allocated even when the branch is not executed?

Reply:
Better like this?